### PR TITLE
Support applicable API during config stage

### PR DESCRIPTION
**Pull request by Machine-Maker**
There are at least 2 events I know of that fire during the configuration stage before the `ServerGamePacketListenerImpl` has been created: `PlayerClientOptionsChangeEvent` and `PlayerSpawnLocationEvent`. In addition, there are the listeners for custom payloads that also run. This PR adds support for the functionality that is supported during the configuration phase, namely kicking players, sending custom resource packs, and sending custom payloads.

```[tasklist]
- [x] kick API
- [x] sending resource pack API
- [x] ~~custom channel stuff~~will handle at a later time
```

### CODE REVIEW

**Comment by MenschenToaster** [comment]
I think we should have an API to keep the player in config for a while as well while e.g. the resource pack is loading. Especially because of this issue right here https://github.com/PaperMC/Velocity/issues/1164.

**Comment by Machine-Maker** [comment]
That is outside the scope of this PR. This is just intending to support already-existing API that is valid for use during the configuration phase. Don't interpret this as saying such API shouldn't exist however.

**Comment by cometcake575** [comment]
Anyone have any idea how long until this PR is merged? Isn't it as simple as just clicking the button now?

**Comment by MiniDigger** [comment]
@cometcake575 this PR is still in draft, it even has a todo list in the description with at least one item that needs looking at.
then the code needs to be reviewed by the team, tested and then it can be integrated. integrating is more complicated than just pressing the button because of how our patching works.
tldr: be patient

File: patches/server/1051-Support-applicable-API-during-config-stage.patch [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 7e6116963d835d4606ef3d93b69d3e44b61288e1..d5a87da4d64651b32e5b1197e932d7722c606978 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -217,6 +217,27 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         this.firstPlayed = System.currentTimeMillis();
+     }
+ 
++    // Paper start - send packets during config stage
++    public void sendPacketDuringMultipleStages(final Packet<? extends net.minecraft.network.protocol.common.ClientCommonPacketListener> packet) {
```
Comment (Lulu13022002) [inline patches/server/1051-Support-applicable-API-during-config-stage.patch]: I don't think this method is needed. Can't you just use commonListener with a null check where needed (i.e the setResourcePack call). Or at least delegate to that method here.

Comment (Machine-Maker) [reply patches/server/1051-Support-applicable-API-during-config-stage.patch]: done

**Comment by Twarug** [comment]
Is there any progress?

**Review by kennytv — APPROVED** [review]

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
In general just want to put this here for more conversation but I feel like we shouldn't "half" support this stage like this. We need to completely deviate from the player inheritance stage in order to properly support this stage... (a disconnect event that occurs here, etc)

Having a player during this stage at all is flawed in itself, and is just due to a bad spigot representation. So we should move away from that in the future anyways.

**Comment by electronicboy** [comment]
I did say elsewhere about making Player more attached to the Network, but, we could maybe go for a more overarching thing of just offering a wrapper for the connection itself which lets you stem off into a configuring client, which lets you see methods only available in that phase, as well as a more general one that lets you get the player? not really sure how that would look, but, we also need to approach this for earlier stage stuff too, anyways, i.e. login plugin messaging API

**Comment by Machine-Maker** [comment]
The Player should be supported in PlayerSpawnLocationEvent which is in a location *after* vanilla creates ServerPlayer but before the game connection is created. So either way, the `Player` interface has to have some support for stuff before the game packet listener is created.

**Comment by Machine-Maker** [comment]
Replaced by configuration phase API.
